File: frontend/src/expr.ts
/** Safe expression evaluator — 1:1 mirror of blender/builders/expr.py.
 * Grammar: numbers, parameter/toggle ids, + - * /, unary minus, parens,
 * min()/max()/abs(). Nothing else, so LLM specs can't execute code. */

const FUNCS: Record<string, (...xs: number[]) => number> = {
  min: Math.min,
  max: Math.max,
  abs: Math.abs,
};

const TOKEN = /\s*(?:([A-Za-z_][A-Za-z0-9_]*)|(\d+\.?\d*|\.\d+)|([()+\-*/,]))/y;

export class ExprError extends Error {}

export function evalExpr(
  input: number | string,
  env: Record<string, number>,
): number {
  if (typeof input === "number") {
    if (!Number.isFinite(input)) throw new ExprError(`Non-finite number`);
    return input;
  }
  if (typeof input !== "string") {
    throw new ExprError(`Expected number or expression, got ${typeof input}`);
  }

  // tokenize
  const tokens: Array<{ kind: "ident" | "num" | "op"; value: string }> = [];
  TOKEN.lastIndex = 0;
  let last = 0;
  let m: RegExpExecArray | null;
  while ((m = TOKEN.exec(input)) !== null) {
    if (m[1]) tokens.push({ kind: "ident", value: m[1] });
    else if (m[2]) tokens.push({ kind: "num", value: m[2] });
    else tokens.push({ kind: "op", value: m[3] });
    last = TOKEN.lastIndex;
  }
  if (last < input.trimEnd().length) {
    throw new ExprError(`Invalid expression ${JSON.stringify(input)}`);
  }

  let pos = 0;
  const peek = () => tokens[pos];
  const take = () => tokens[pos++];
  const expectOp = (op: string) => {
    const t = take();
    if (!t || t.kind !== "op" || t.value !== op) {
      throw new ExprError(`Expected '${op}' in ${JSON.stringify(input)}`);
    }
  };

  function parseExpr(): number {
    let left = parseTerm();
    while (peek()?.kind === "op" && (peek().value === "+" || peek().value === "-")) {
      const op = take().value;
      const right = parseTerm();
      left = op === "+" ? left + right : left - right;
    }
    return left;
  }

  function parseTerm(): number {
    let left = parseFactor();
    while (peek()?.kind === "op" && (peek().value === "*" || peek().value === "/")) {
      const op = take().value;
      const right = parseFactor();
      left = op === "*" ? left * right : left / right;
    }
    return left;
  }

  function parseFactor(): number {
    const t = take();
    if (!t) throw new ExprError(`Unexpected end of expression ${JSON.stringify(input)}`);
    if (t.kind === "num") return parseFloat(t.value);
    if (t.kind === "op" && t.value === "-") return -parseFactor();
    if (t.kind === "op" && t.value === "+") return parseFactor();
    if (t.kind === "op" && t.value === "(") {
      const v = parseExpr();
      expectOp(")");
      return v;
    }
    if (t.kind === "ident") {
      if (peek()?.kind === "op" && peek().value === "(") {
        const fn = FUNCS[t.value];
        if (!fn) throw new ExprError(`Unknown function ${t.value}()`);
        take(); // (
        const args = [parseExpr()];
        while (peek()?.kind === "op" && peek().value === ",") {
          take();
          args.push(parseExpr());
        }
        expectOp(")");
        return fn(...args);
      }
      const v = env[t.value];
      if (v === undefined) {
        throw new ExprError(
          `Unknown name '${t.value}' (known: ${Object.keys(env).join(", ") || "<none>"})`,
        );
      }
      return v;
    }
    throw new ExprError(`Unexpected token '${t.value}' in ${JSON.stringify(input)}`);
  }

  const result = parseExpr();
  if (pos !== tokens.length) {
    throw new ExprError(`Trailing tokens in ${JSON.stringify(input)}`);
  }
  if (!Number.isFinite(result)) {
    throw new ExprError(`Expression ${JSON.stringify(input)} is not finite`);
  }
  return result;
}

File: blender/builders/expr.py
"""Safe arithmetic-expression evaluator for custom-primitive specs.

The LLM's "generate anything" path emits primitive dimensions as expressions
over parameter/toggle ids ("seat_height + 0.02"). This evaluator accepts ONLY
numbers, known names, + - * /, unary minus, parentheses, and min/max/abs —
anything else (attributes, subscripts, arbitrary calls) is rejected, so LLM
output can never execute code. Mirrored 1:1 in frontend/src/expr.ts.
"""
from __future__ import annotations

import ast
import math
import operator
from typing import Mapping, Union

Number = Union[int, float]

_BIN_OPS = {
    ast.Add: operator.add,
    ast.Sub: operator.sub,
    ast.Mult: operator.mul,
    ast.Div: operator.truediv,
}

_FUNCS = {"min": min, "max": max, "abs": abs}


class ExprError(ValueError):
    pass


def _eval(node: ast.AST, env: Mapping[str, float]) -> float:
    if isinstance(node, ast.Constant):
        if isinstance(node.value, (int, float)) and not isinstance(node.value, bool):
            return float(node.value)
        raise ExprError(f"Non-numeric constant: {node.value!r}")
    if isinstance(node, ast.Name):
        try:
            return float(env[node.id])
        except KeyError:
            raise ExprError(
                f"Unknown name {node.id!r} (known: {', '.join(sorted(env)) or '<none>'})"
            ) from None
    if isinstance(node, ast.BinOp) and type(node.op) in _BIN_OPS:
        left, right = _eval(node.left, env), _eval(node.right, env)
        try:
            return _BIN_OPS[type(node.op)](left, right)
        except ZeroDivisionError:
            raise ExprError(f"Division by zero: {ast.dump(node)}") from None
    if isinstance(node, ast.UnaryOp) and isinstance(node.op, (ast.USub, ast.UAdd)):
        v = _eval(node.operand, env)
        return -v if isinstance(node.op, ast.USub) else v
    if (
        isinstance(node, ast.Call)
        and isinstance(node.func, ast.Name)
        and node.func.id in _FUNCS
        and not node.keywords
        and node.args
    ):
        return float(_FUNCS[node.func.id](*(_eval(a, env) for a in node.args)))
    raise ExprError(f"Disallowed expression element: {ast.dump(node)}")


def safe_eval(expr: Union[str, Number], env: Mapping[str, float]) -> float:
    """Evaluate a number or an expression string against ``env``."""
    if isinstance(expr, (int, float)) and not isinstance(expr, bool):
        if not math.isfinite(expr):
            raise ExprError(f"Non-finite number: {expr!r}")
        return float(expr)
    if not isinstance(expr, str):
        raise ExprError(f"Expected number or expression string, got {type(expr).__name__}")
    try:
        tree = ast.parse(expr, mode="eval")
    except SyntaxError as exc:
        raise ExprError(f"Invalid expression {expr!r}: {exc.msg}") from None
    result = _eval(tree.body, env)
    if not math.isfinite(result):
        raise ExprError(f"Expression {expr!r} is not finite")
    return result
